README
WWW-Selenium-Utils
==================

This package contains useful helper libraries and scripts for 
testing websites using Selenium.

  http://selenium.thoughtworks.com

The selenium-regen script can be used to automatically convert test scripts
written in a simple wiki-like table markup into the html tables that Selenium
expects.

Sample .wiki file:

  Test Title
  | open | /foo |
  | verifyLocation | /foo | 
  | verifyTextPresent | id=name | Luke |
  # comments are allowed

  # so are blank lines
  | clickAndWait | id=main_btn |

This package contains two modules:

  WWW::Selenium::Utils
    generate_suite() - converts wiki files, creates TestSuite.html

  WWW::Selenium::Utils::CGI
    run() - runs a command, returns the output
    cat() - dump the contents of a file
    state() - query and store simple key/value pairs to keep state

    You will need to wrap these functions in CGI or mod_perl handlers.
    Example scripts can be found in the scripts/ directory.
    
And some example scripts:

  script/postResults.cgi - writes a Selenium result post to disk

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

None.

THANKS TO

Mark Stosberg for early feedback.  Tom Milne for inspiration for state().

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2005 Luke Closs

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 



Changes
Revision history for Perl extension WWW-Selenium-Utils.

0.06 - Mon Dec 12 00:22:25 PST 2005
    - added --base option to selenium-regen to prefix a common
      path to all open and assert/verifyLocation commands
      - Thanks Mark Stosberg!
    - reword the auto-generated comment
    - bugfixes for directory paths

0.05 - Sat Dec 10 15:42:26 PST 2005
    - added Auto-generated comment to generated html files
      (thanks Mark Stosberg!)
    - added PostResults.pm to write selenium results to file
    - added script/postResults.cgi   

0.04 - Mon Nov 28 23:09:42 PST 2005
    - Added state functionality
    - changed --selenium to default parameter
    - selenium-regen needs the tests dir, not the selenium dir
      - Thanks Mark Stosberg!
    - use test title in TestSuite.html (better titles)

0.03  Later
    - moved selenium-regen to bin from script

0.01  Fri Nov  4 10:05:50 2005
	- original version; created by h2xs 1.21 with options
		-AXP -b 5.6.0 WWW-Selenium-Utils



t/Regen.pm
package t::Regen;
use strict;
use warnings;
use base 'Exporter';
our @EXPORT_OK = qw(test_setup);
use File::Path;

sub test_setup {
    my $kind = shift || '';
    my $testdir = "t/tests";
    !-d $testdir or rmtree $testdir or die "Can't rmtree $testdir: $!";
    mkpath $testdir or die "Can't mkpath $testdir: $!";
    open(my $fh, ">$testdir/foo.wiki") or die "Can't open $testdir/foo.wiki: $!";
    my $invalid = '';
    $invalid = "nintendo | sony | xbox |" if $kind eq 'invalid line';
    print $fh <<EOT;
    some title
    | open | /foo |
    | verifyText | id=foo | bar |
# comment
$invalid
# next line has spaces at the end
    | verifyLocation | /bar |   
EOT
    close $fh or die "Can't write $testdir/foo.wiki: $!";

    if ($kind eq "with orphan") {
        open($fh, ">$testdir/orphan.html") or 
            die "Can't open $testdir/orphan.html: $!";
        print $fh "Auto-generated from $testdir/orphan.wiki at 23";
        close $fh or die "Can't write $testdir/orphan.wiki: $!";
    }

    open($fh, ">$testdir/bar.html") or die "Can't open $testdir/bar.html: $!";
    print $fh <<EOT;
    <html>
      <body>
        <table>
          <tr>
            <td>Test title</td>
          </tr>
          <tr>
            <td>open</td><td>/foo</td><td></td>
          </tr>
        </table>
      </body>
    </html>
EOT
    close $fh or die "Can't write $testdir/bar.html: $!";
    return $testdir;
}

1;


lib/WWW/Selenium/Utils/CGI.pm
package WWW::Selenium::Utils::CGI;
use 5.006;
use strict;
use warnings;
use Carp;
use File::Find;
use CGI qw(:standard);
use Config;
use Data::Dumper;

require Exporter;
our @ISA = qw(Exporter);
our @EXPORT_OK = qw(run cat state);

sub run {
    my $q = shift or croak("CGI query object is mandatory!");

    my $cmd = $q->param('cmd');
    return error("cmd is a mandatory parameter!") unless $cmd;

    my $results = qx($cmd) || $!;
    $results =~ s/</&lt;/g;
    return header . start_html("Output of \"$cmd\"") 
           . h1("Output of \"$cmd\":") . pre($results) . end_html;
}


sub cat {
    my $q = shift or croak("CGI query object is mandatory!");
    my %opts = @_;
    my $basedir = $opts{basedir} || $Config{prefix};

    my $file = $q->param('file');
    my $raw  = $q->param('raw');

    return error("file is a mandatory parameter!") unless $file;

    $file = "$basedir/$file" unless $file =~ m#^/#;
    return error("Sorry, $file doesn't exist!") unless -e $file;

    my $contents;
    open(my $fh, $file) or return error("Can't open $file: $!");
    { 
        local $/ = undef;
        $contents = <$fh>;
    }
    close $fh or return error("Can't close $file: $!");

    return header . $contents if $raw;

    $contents =~ s/</&lt;/g;
    $contents = pre($contents);
    return header . start_html("Contents of $file") . $contents . end_html;
}

sub state {
    my $q = shift or croak("CGI query object is mandatory!");

    if ($q->param('clear_state')) {
        _clear_state();
        return header . start_html . h1("State cleared");
    }

    my $key = $q->param('key');
    my $value = $q->param('value');
    return error('key is a mandatory parameter!') unless $key;
    unless ($value) {
        my $val = _read_state($key);
        return error("'$key' is not a valid key!") unless defined $val;
        return header . start_html . h1("State for $key")
               . "'$key' is '$val'";
    }
    eval {
        _store_state( $key => $value );
    };
    if ($@) {
        return header . start_html . h1("State for $key")
               . "Error saving key $key: $@";
    }
    return header . start_html . h1("State for $key")
           . "Stored '$value' in '$key'";
}

my $statefile = "/tmp/selenium-utils-state";

sub _read_state {
    my $key = shift;
    my $content = '';
    if (-e $statefile) {
        local $/;
        open(my $fh, $statefile) or die "Can't open $statefile: $!";
        $content = <$fh>;
        close $fh or die "Can't close $statefile: $!";
    }

    my $state;
    eval $content;
    $state ||= {};
    return $key ? $state->{$key} : $state;
}

sub _store_state {
    my ($key, $val) = @_;
    my $state = _read_state;
    $state->{$key} = $val;
    my $tmpstate = "$statefile.$$";
    my $textstate = Data::Dumper->Dump([$state], ["state"]);
    open(my $fh, ">$tmpstate") or die "Can't open $tmpstate: $!";
    print $fh $textstate;
    close $fh or die "Can't write $tmpstate: $!";
    rename $tmpstate, $statefile or 
        die "Can't rename $tmpstate, $statefile: $!";
}

sub _clear_state {
    unlink $statefile;
}

sub error {
    my $msg = shift;
    return header . start_html . h1("Error!") . $msg . end_html;
}

1;
__END__

=head1 NAME

WWW::Selenium::Utils::CGI - helper functions Selenium CGIs

=head1 SYNOPSIS

  # simple CGI script
  use WWW::Selenium::Utils::CGI qw(run);
  use CGI;
  print run( CGI->new() );

=head1 DESCRIPTION

This package contains useful functions for creating test scaffolding
CGI scripts that Selenium can use.  Users of this module will need
to create their own cgi or mod_perl handlers that call these functions.

=head1 SUBROUTINES

These subs should be called by CGI or mod_perl wrappers.  I feel
this will provide the most generic interface that can be customized
to any application.  Sample CGI scripts are in scripts/ of this 
distribution.  Indeed I use both CGI and mod_perl for various test 
projects.

=head2 run

C<run()> will run an command passed in from as a CGI variable.

Note that this is NOT SAFE for general websites, as it allows 
ARBITRARY COMMANDS to be run.  For testing purposes however, it 
is quite useful.

HTTP GET Parameters:

=over 4

=item cmd

The command to execute.  Note that you will need to properly encode
commands in your selenium test cases, like this:

  /open /selenium/run.cgi?cmd=ls%20-l

=back

=head2 cat

C<cat()> will output the contents of a file.

Arguments:

=over 4

=item basedir

The base directory when a relative path is used.  Defaults to
perl's install prefix.

=back

HTTP GET Parameters:

=over 4

=item file

The file to read.  If the file does not begin with a '/', then the file
will be relative to $Config{prefix} (where your perl is installed to).

=item raw

If this is false (the default), then the contents of the file will be
surrownded in a pre block, to make the output look nicer in the browser.

=back

=head2 state

Keeps any kind of key/value state.  This can allow selenium to signal
other processes or stare information.  State is stored in a flat file
in /tmp.

HTTP GET Parameters:

=over 4

=item key (Mandatory)

A key to set or get.

=item value

If present, will be stored for the given key.

=back

=head1 DIAGNOSTICS

None.

=head1 DEPENDENCIES

Uses CGI.

=head1 BUGS AND LIMITATIONS

There are no known bugs in this module.
Please report problums to Luke Closs (cpan@5thplane.com).
Patches are welcome.

=head1 AUTHOR

Luke Closs (cpan@5thplane.com)

=head1 LICENCE AND COPYRIGHT

Copyright (c) 2005 Luke Closs (cpan@5thplane.com).  All rights reserved.

This module is free software; you can redstribute it and/or
modify it under the same terms as Perl itself.  See L<perlartistic>.

This program is distributed in the hope that it will be useful, 
but WITHOUT ANY WARRANTY; without even the implied warranty of 
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.



lib/WWW/Selenium/Utils.pm
package WWW::Selenium::Utils;

use 5.006;
use strict;
use warnings;
use Carp;
use File::Find;

require Exporter;
our @ISA = qw(Exporter);
our @EXPORT_OK = qw(generate_suite cat);

our $VERSION = '0.06';

sub html_header;
sub html_footer;

sub generate_suite {
    my %opts = @_;

    my $testdir = $opts{test_dir} or croak("test_dir is mandatory!");
    croak("$testdir is not a directory!") unless -d $testdir;
    $testdir =~ s#/$##;
    my $files = $opts{files} || test_files($testdir);
    my $verbose = $opts{verbose};

    my $suite = "$testdir/TestSuite.html";
    my $date = localtime;

    open(my $fh, ">$suite.tmp") or die "Can't open $suite.tmp: $!";
    print $fh html_header(title => "Test Suite",
                          text  => "Generated at $date",
                         );

    for (sort {$a cmp $b} @$files) {
        next if /(?:\.tmp|TestSuite\.html)$/;

        my $f = $_;
        my $fp = "$testdir/$f";
        if ($f =~ /(.+)\.html$/) {
            my $basename = $1;
            # skip html files that we have or will generate
            next if -e "$testdir/$basename.wiki";
            # find orphaned html files
            my $html = cat($fp);
            if ($html =~ m#Auto-generated from $testdir/$basename\.wiki# and 
                        !-e "$testdir/$basename.wiki") {
                print "Deleting orphaned file $fp\n" if $verbose;
                unlink $fp or die "Can't unlink $fp: $!";
                next;
            }
        }

        print "Adding row for $f\n" if $verbose;
        if (/\.wiki$/) {
            $f = wiki2html($fp, 
                           verbose => $verbose,
                           base_href => $opts{base_href});
            $f =~ s/^$testdir\///;
            $fp = "$testdir/$f";
        }
        my $title = find_title($fp);
        print $fh qq(\t<tr><td><a href="./$f">$title</a></td></tr>\n);
    }
    #print the footer
    print $fh html_footer();

    # save and rename into place
    close $fh or die "Can't close $suite.tmp: $!";
    rename "$suite.tmp", $suite or die "can't rename $suite.tmp $suite: $!";
    print "Created new $suite\n" if $verbose;
}

sub test_files {
    my $testdir = shift;

    my @tests;
    find(sub {
            my $n = $File::Find::name;
            return if -d $n;
            return unless m#(?:wiki|html)$#;
            $n =~ s#^$testdir/?##;
            $n =~ s#^.+/tests/##;
            push @tests, $n;
        }, $testdir);
    return \@tests;
}

sub wiki2html {
    my ($wiki, %opts) = @_;
    my $verbose = $opts{verbose};
    my $base_href = $opts{base_href};
    $base_href =~ s#/$## if $base_href;

    (my $html = $wiki) =~ s#\.wiki$#.html#;

    open(my $in, $wiki) or die "Can't open $wiki: $!";
    my $title = <$in>;
    chomp $title;
    $title =~ s#^\s*##;
    $title =~ s#^\|(.+)\|$#$1#;
    print "Generating html for ($title): $html\n" if $verbose;

    open(my $out, ">$html") or die "Can't open $html: $!";
    my $now = localtime;
    print $out html_header( title => $title,
                            text => "<b>Auto-generated from $wiki</b><br />");
    while(<$in>) {
        s/^\s*//;
        next if /^#/ or /^\s*$/;
        chomp;
        if (/^\s*\|\s*(.+?)\s*\|\s*$/) {
            my ($cmd, $opt1, $opt2) = split /\s*\|\s*/, $1, 3;
            die "No command found! ($_)" unless $cmd;
            $opt1 ||= '&nbsp;';
            $opt2 ||= '&nbsp;';
            if ($base_href and ($cmd eq "open" or 
                                $cmd =~ /(?:assert|verify)Location/)) {
                $opt1 =~ s#^/##;
                $opt1 = "$base_href/$opt1";
            }
            print $out "\n\t<tr><td>$cmd</td>"
                       . "\n\t<td>$opt1</td>"
                       . "\n\t<td>$opt2</td></tr>\n";
        }
        else {
            warn "Invalid line ($.) in file $wiki: $_\n";
        }
    }
    close $in or die "Can't close $wiki: $!";

    print $out html_footer("<hr />Auto-generated from $wiki at $now\n");
    close $out or die "Can't write $html: $!";
    return $html;
}

sub find_title {
    my $filename = shift;

    open(my $fh, $filename) or die "Can't open $filename: $!";
    my $contents;
    { 
        local $/;
        $contents = <$fh>;
    }
    close $fh or die "Can't close $filename: $!";

    return $filename unless $contents;
    return $1 if $contents =~ m#<title>\s*(.+)\s*</title>#;
    return $1 if $filename =~ m#^.+/(.+)\.html$#;
    return $filename;
}

sub html_header {
    my %opts = @_;
    my $title = $opts{title} || 'Generic Title';
    my $text = $opts{text} || '';

    my $header = <<EOT;
<html>
  <head>
    <meta content="text/html; charset=ISO-8859-1"
          http-equiv="content-type">
    <title>$title</title>
  </head>
  <body>
    $text
    <table cellpadding="1" cellspacing="1" border="1">
      <tbody>
        <tr>
          <td rowspan="1" colspan="3">$title</td>
        </tr>
EOT
    return $header;
}

sub html_footer {
    my $text = shift || '';
    return <<EOT;
      </tbody>
    </table>
    $text
  </body>
</html>
EOT
}

sub cat {
    my $file = shift;
    my $contents;
    eval {
        open(my $fh, $file) or die "Can't open $file: $!";
        { 
            local $/;
            $contents = <$fh>;
        }
        close $fh or die "Can't close $file: $!";
    };
    warn if $@;
    return $contents;
}


1;
__END__

=head1 NAME

WWW::Selenium::Utils - helper functions for working with Selenium

=head1 SYNOPSIS

  use WWW::Selenium::Utils qw(generate_suite);

  # convert .wiki files to .html and create TestSuite.html
  generate_suite( test_dir => "/var/www/selenium/tests",
                  base_href => "/monkey",
                  verbose => 1,
                );

=head1 DESCRIPTION

This package contains utility functions for working with Selenium.

=head1 SUBROUTINES

=head2 generate_suite

C<generate_suite()> will convert all .wiki files in selenium/tests to .html,
and then create a TestSuite.html file that contains links to all the .html 
files.

The .wiki files are much easier to read and write.  The format of .wiki files
is like this:

  title
  | cmd | opt1 | opt2 |
  | cmd | opt1 |
  # comment

  # empty lines are ignored
  # comments are ignored too

Parameters:

=over 4

=item test_dir

The path to the 'tests' directory inside selenium.

=item verbose

If true, informative messages will be printed.

=item base_href

Will prepend the given location to all locations for the
open and assert/verifyLocation commands.

=back

=head1 DIAGNOSTICS

If you set the C<verbose> option to 1 when calling generate_suite, the function
will print lines detailing what it is doing.

=head1 DEPENDENCIES

Uses CGI.

=head1 BUGS AND LIMITATIONS

There are no known bugs in this module.
Please report problums to Luke Closs (cpan@5thplane.com).
Patches are welcome.

=head1 AUTHOR

Luke Closs (cpan@5thplane.com)

=head1 LICENCE AND COPYRIGHT

Copyright (c) 2005 Luke Closs (cpan@5thplane.com).  All rights reserved.

This module is free software; you can redstribute it and/or
modify it under the same terms as Perl itself.  See L<perlartistic>.

This program is distributed in the hope that it will be useful, 
but WITHOUT ANY WARRANTY; without even the implied warranty of 
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

